libqpdf/QPDFWriter.cc

```diff
@@ -1274,8 +1274,14 @@ QPDFWriter::willFilterStream(
         QPDFObjectHandle filter_obj = stream_dict.getKey("/Filter");
         if (!m->recompress_flate && !stream.isDataModified() && filter_obj.isName() &&
             (filter_obj.getName() == "/FlateDecode" || filter_obj.getName() == "/Fl")) {
-            QTC::TC("qpdf", "QPDFWriter not recompressing /FlateDecode");
-            filter = false;
+            bool empty_stream = stream_dict.getKey("/Length").getIntValue() == 0;
+            if (empty_stream) {
+                QTC::TC("qpdf", "QPDFWriter is removing /FlateDecode from an empty stream");
+                filter = true;
```

Comment:
If you add a new QTC::TC entry, you also have to update the qpdf/qpdf.testcov file and ensure that there is a test case to exercise the line during CI. See 
[coverage](https://qpdf.readthedocs.io/en/stable/contributing.html#coverage) in the manual.